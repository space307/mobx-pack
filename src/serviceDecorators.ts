import { cloneDeep, last } from 'lodash';
import type {
  BindableEntity,
  BindableEntityConstructor,
  BinderConfig,
  Constructor,
  ServiceConfigCallbackSet,
} from './typing/common.js';

function validateName(name: string): boolean {
  return !!(name && typeof name === 'string' && /^[A-Za-z][A-Za-z0-9_]+$/.test(name));
}

function validateNameList(list: string[]): boolean {
  return list.reduce((acc, item) => {
    if (!validateName(item)) {
      acc = false;
    }
    return acc;
  }, true);
}

function createConfig(): BinderConfig {
  return {
    onStart: [],
    onStop: '',
    bindAs: '',
    onBind: [],
    onUnbind: [],
  };
}

function prepareConfig<T extends Constructor>(ctr: T): T & BindableEntityConstructor {
  const bindableCtr = ctr as unknown as T & BindableEntityConstructor;

  if (!bindableCtr.binderConfig) {
    bindableCtr.binderConfig = createConfig();
  } else {
    bindableCtr.binderConfig = cloneDeep(bindableCtr.binderConfig);
  }

  return bindableCtr;
}

function putServiceNamesToConfig(
  names: string[],
  ctr: BindableEntityConstructor,
  callbackName: string,
  optionName: keyof Pick<BinderConfig, 'onStart' | 'onBind' | 'onUnbind'>,
): void {
  if (names && names.length && callbackName) {
    names.forEach((serviceName: string) => {
      if (!validateName(serviceName)) {
        throw new Error(`Wrong service name "${serviceName}"
          passed to function "${callbackName}" decorator (service:${ctr.name ?? 'unknown'}).`);
      }
    });

    if (!ctr.binderConfig) {
      throw new Error(`binderConfig not found! (service:${ctr.name ?? 'unknown'}).`);
    }

    if (optionName === 'onBind' || optionName === 'onUnbind') {
      const option = ctr.binderConfig[optionName];

      if (option) {
        const existCallback = option.find(
          (callback: ServiceConfigCallbackSet): boolean => last(callback) === callbackName,
        );

        if (existCallback === undefined) {
          option.push([...names, callbackName]);
        } else {
          existCallback.splice(0, existCallback.length, ...names, callbackName);
        }
      }
    } else {
      ctr.binderConfig[optionName] = [...names, callbackName];
    }
  }
}

function putMethodNameToConfig(
  ctr: BindableEntityConstructor,
  callbackName: string,
  optionName: keyof Pick<BinderConfig, 'onStop'>,
): void {
  if (ctr.binderConfig) {
    ctr.binderConfig[optionName] = callbackName;
  }
}

export function bindAs(
  name: string,
): <T extends object>(ctr: Constructor<T>) => BindableEntityConstructor<T & BindableEntity> {
  if (typeof name === 'function') {
    throw new Error(
      `Wrong attributes passed to bindAs decorator (service:${(name as () => unknown).name}).`,
    );
  }
  return ctr => {
    if (!validateName(name)) {
      throw new Error(
        `Wrong name "${name}" passed to bindAs decorator (service:${ctr.name ?? 'unknown'}).`,
      );
    }

    const bindedCtr = prepareConfig(ctr);
    bindedCtr.binderConfig.bindAs = name;

    return bindedCtr;
  };
}

export function onBind<T extends object>(
  ...serviceNames: string[]
): (service: T, callbackName: string) => void {
  if (!serviceNames.length || !validateNameList(serviceNames)) {
    throw new Error(`Wrong attributes passed to onBind decorator (${serviceNames.join(',')}).`);
  }

  return (service, callbackName) => {
    const proto = service.constructor as BindableEntityConstructor;
    prepareConfig(proto);

    putServiceNamesToConfig(serviceNames, proto, callbackName, 'onBind');
  };
}

export function onUnbind<T extends object>(
  ...serviceNames: string[]
): (service: T, callbackName: string) => T {
  if (!serviceNames.length || !validateNameList(serviceNames)) {
    throw new Error(`Wrong attributes passed to onUnbind decorator (${serviceNames.join(',')}).`);
  }
  return (service, callbackName) => {
    const proto = service.constructor as BindableEntityConstructor;
    prepareConfig(proto);

    putServiceNamesToConfig(serviceNames, proto, callbackName, 'onUnbind');
    return service;
  };
}

export function onStart<T extends object>(
  ...serviceNames: string[]
): (service: T, callbackName: string) => T {
  if (!serviceNames.length || !validateNameList(serviceNames)) {
    throw new Error(`Wrong attributes passed to onStart decorator (${serviceNames.join(',')}).`);
  }
  return (service, callbackName) => {
    const proto = service.constructor as BindableEntityConstructor;
    prepareConfig(proto);

    putServiceNamesToConfig(serviceNames, proto, callbackName, 'onStart');
    return service;
  };
}

export function onStop<T extends object>(service: T, callbackName: string): void {
  const proto = service.constructor as BindableEntityConstructor;
  prepareConfig(proto);
  putMethodNameToConfig(proto, callbackName, 'onStop');
}
