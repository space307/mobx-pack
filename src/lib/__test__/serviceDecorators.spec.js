/* eslint-disable no-unused-vars */
import { onStop, onStart, unbindServices, bindServices, bindAs } from '../serviceDecorators.js';


describe('serviceDecorators test', () => {
  it('wrong bindAs param', () => {
    expect(() => {
      @bindAs
      class Service {
      }
    }).toThrow();

    expect(() => {
      @bindAs()
      class Service {
      }
    }).toThrow();

    expect(() => {
      @bindAs(1)
      class Service {
      }
    }).toThrow();
  });

  it('wrong bindServices param', () => {
    expect(() => {
      @bindAs('test')
      class Service {
        @bindServices
        onBind() {}
      }
    }).toThrow();

    expect(() => {
      @bindAs('test')
      class Service {
        @bindServices()
        onBind() {}
      }
    }).toThrow();
    expect(() => {
      @bindAs('test')
      class Service {
        @bindServices('test1', 1)
        onBind() {}
      }
    }).toThrow();
  });


  it('serviceDecorators config', () => {
    @bindAs('test')
    class Service {
      @onStart
      onStart() {}

      @bindServices('test1', 'test2')
      onBind() {}

      @bindServices('test1', 'test2')
      onBindOnlyGarageStore() {}

      @unbindServices('test1', 'test2')
      onUnbind() {}

      @onStop
      onStop() {}
    }


    expect(Service.binderConfig).toMatchSnapshot();
  });
});
