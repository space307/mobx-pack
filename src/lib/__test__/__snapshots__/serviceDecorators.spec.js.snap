// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serviceDecorators test serviceDecorators config 1`] = `
Object {
  "config": Object {
    "bindAs": "test",
    "onBind": Array [
      Array [
        "test1",
        "test2",
        "onBind",
      ],
      Array [
        "test1",
        "test2",
        "onBindOnlyGarageStore",
      ],
    ],
    "onUnbind": Array [
      Array [
        "test1",
        "test2",
        "onUnbind",
      ],
    ],
  },
  "onStart": "onStart",
  "onStop": "onStop",
}
`;
